File: mobile/android/app/src/main/kotlin/com/xiaozhi/network/NetworkManager.kt
package com.xiaozhi.network

import android.content.Context
import android.util.Base64
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

object NetworkManager {
    
    private const val BASE_URL_DEBUG = "http://10.0.2.2:5000"  // Android emulator localhost
    private const val BASE_URL_PROD = "https://your-domain.replit.app"
    private const val WS_URL_DEBUG = "ws://10.0.2.2:5000/ws/mobile"
    private const val WS_URL_PROD = "wss://your-domain.replit.app/ws/mobile"
    
    private val isDebug = true  // BuildConfig.DEBUG
    
    val baseUrl: String get() = if (isDebug) BASE_URL_DEBUG else BASE_URL_PROD
    private val wsUrl: String get() = if (isDebug) WS_URL_DEBUG else WS_URL_PROD
    
    private lateinit var client: OkHttpClient
    private var webSocket: WebSocket? = null
    private lateinit var deviceId: String
    
    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected
    
    private val _connectionError = MutableStateFlow<String?>(null)
    val connectionError: StateFlow<String?> = _connectionError
    
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var pingJob: Job? = null
    
    var onASRResult: ((String, Boolean) -> Unit)? = null
    var onASRInterim: ((String) -> Unit)? = null
    var onAIReplyChunk: ((String, Boolean) -> Unit)? = null
    var onReminder: ((String, String) -> Unit)? = null
    var onSyncRequired: (() -> Unit)? = null
    
    fun initialize(context: Context) {
        deviceId = DeviceManager.getDeviceId(context)
        
        client = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(60, TimeUnit.SECONDS)
            .pingInterval(30, TimeUnit.SECONDS)
            .build()
    }
    
    fun connect() {
        val url = "$wsUrl?deviceId=$deviceId&role=MASTER&platform=Android"
        val request = Request.Builder().url(url).build()
        
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _isConnected.value = true
                _connectionError.value = null
                startPingTimer()
            }
            
            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }
            
            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                _isConnected.value = false
                stopPingTimer()
            }
            
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                _isConnected.value = false
                _connectionError.value = t.message
                stopPingTimer()
                
                scope.launch {
                    delay(5000)
                    connect()
                }
            }
        })
    }
    
    fun disconnect() {
        webSocket?.close(1000, "User disconnect")
        webSocket = null
        _isConnected.value = false
        stopPingTimer()
    }
    
    private fun handleMessage(text: String) {
        try {
            val json = JSONObject(text)
            val type = json.optString("type")
            
            when (type) {
                "pong" -> { }
                "asr_result" -> {
                    val resultText = json.optString("text")
                    val isFinal = json.optBoolean("isFinal", true)
                    onASRResult?.invoke(resultText, isFinal)
                }
                "asr_interim" -> {
                    val interimText = json.optString("text")
                    onASRInterim?.invoke(interimText)
                }
                "ai_reply_chunk" -> {
                    val chunkText = json.optString("text")
                    val isFinal = json.optBoolean("isFinal", false)
                    onAIReplyChunk?.invoke(chunkText, isFinal)
                }
                "reminder" -> {
                    val data = json.optJSONObject("data")
                    val title = data?.optString("title") ?: "提醒"
                    val body = data?.optString("body") ?: ""
                    onReminder?.invoke(title, body)
                }
                "sync_required" -> {
                    onSyncRequired?.invoke()
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
    
    fun send(message: JSONObject) {
        webSocket?.send(message.toString())
    }
    
    fun sendAudioChunk(audioData: ByteArray, sequence: Int, isFinal: Boolean) {
        val message = JSONObject().apply {
            put("type", "audio_chunk")
            put("data", Base64.encodeToString(audioData, Base64.NO_WRAP))
            put("sequence", sequence)
            put("final", isFinal)
        }
        send(message)
    }
    
    fun startASR() {
        send(JSONObject().apply {
            put("type", "asr_start")
            put("format", "pcm")
            put("sampleRate", 16000)
        })
    }
    
    fun stopASR() {
        send(JSONObject().apply { put("type", "asr_stop") })
    }
    
    private fun startPingTimer() {
        pingJob = scope.launch {
            while (isActive) {
                delay(30_000)
                send(JSONObject().apply { put("type", "ping") })
            }
        }
    }
    
    private fun stopPingTimer() {
        pingJob?.cancel()
        pingJob = null
    }
    
    suspend fun <T> request(
        endpoint: String,
        method: String = "GET",
        body: JSONObject? = null,
        parser: (String) -> T
    ): Result<T> = withContext(Dispatchers.IO) {
        try {
            val requestBuilder = Request.Builder()
                .url("$baseUrl$endpoint")
                .header("Content-Type", "application/json")
                .header("X-User-Role", "MASTER")
                .header("X-Device-Id", deviceId)
                .header("X-Platform", "Android")
                .header("X-App-Version", "1.0.0")
            
            when (method) {
                "POST" -> {
                    val requestBody = (body?.toString() ?: "{}").toRequestBody("application/json".toMediaType())
                    requestBuilder.post(requestBody)
                }
                "PUT" -> {
                    val requestBody = (body?.toString() ?: "{}").toRequestBody("application/json".toMediaType())
                    requestBuilder.put(requestBody)
                }
                "DELETE" -> requestBuilder.delete()
            }
            
            val response = client.newCall(requestBuilder.build()).execute()
            val responseBody = response.body?.string() ?: ""
            
            if (response.isSuccessful) {
                Result.success(parser(responseBody))
            } else {
                val error = try {
                    JSONObject(responseBody).optString("error", "请求失败")
                } catch (e: Exception) {
                    "请求失败: ${response.code}"
                }
                Result.failure(Exception(error))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun chat(message: String, sessionId: String? = null): Result<ChatResponse> {
        val body = JSONObject().apply {
            put("message", message)
            sessionId?.let { put("sessionId", it) }
        }
        
        return request("/api/conversation/chat", "POST", body) { responseText ->
            val json = JSONObject(responseText)
            ChatResponse(
                reply = json.optString("reply"),
                sessionId = json.optString("sessionId"),
                intent = json.optString("intent"),
                confidence = json.optDouble("confidence", 0.0)
            )
        }
    }
}

data class ChatResponse(
    val reply: String,
    val sessionId: String,
    val intent: String?,
    val confidence: Double
)

File: mobile/ios/XiaoZhi/Services/NetworkManager.swift
import Foundation
import Combine

class NetworkManager: NSObject, ObservableObject {
    static let shared = NetworkManager()
    
    @Published var isConnected = false
    @Published var connectionError: String?
    
    private var webSocket: URLSessionWebSocketTask?
    private var session: URLSession!
    private var cancellables = Set<AnyCancellable>()
    private var pingTimer: Timer?
    
    private let baseURL: String
    private let wsURL: String
    
    override init() {
        #if DEBUG
        self.baseURL = "http://localhost:5000"
        self.wsURL = "ws://localhost:5000/ws/mobile"
        #else
        self.baseURL = "https://your-domain.replit.app"
        self.wsURL = "wss://your-domain.replit.app/ws/mobile"
        #endif
        
        super.init()
        self.session = URLSession(configuration: .default, delegate: self, delegateQueue: .main)
    }
    
    func connect() {
        guard let url = URL(string: "\(wsURL)?deviceId=\(DeviceManager.deviceId)&role=MASTER&platform=iOS") else {
            return
        }
        
        webSocket = session.webSocketTask(with: url)
        webSocket?.resume()
        receiveMessage()
        startPingTimer()
    }
    
    func disconnect() {
        webSocket?.cancel(with: .normalClosure, reason: nil)
        webSocket = nil
        isConnected = false
        stopPingTimer()
    }
    
    private func receiveMessage() {
        webSocket?.receive { [weak self] result in
            switch result {
            case .success(let message):
                self?.handleMessage(message)
                self?.receiveMessage()
            case .failure(let error):
                self?.connectionError = error.localizedDescription
                self?.isConnected = false
            }
        }
    }
    
    private func handleMessage(_ message: URLSessionWebSocketTask.Message) {
        switch message {
        case .string(let text):
            parseServerMessage(text)
        case .data(let data):
            if let text = String(data: data, encoding: .utf8) {
                parseServerMessage(text)
            }
        @unknown default:
            break
        }
    }
    
    private func parseServerMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else {
            return
        }
        
        switch type {
        case "pong":
            break
        case "asr_result":
            handleASRResult(json)
        case "asr_interim":
            handleASRInterim(json)
        case "ai_reply_chunk":
            handleAIReplyChunk(json)
        case "reminder":
            handleReminder(json)
        case "sync_required":
            handleSyncRequired(json)
        default:
            print("[WS] Unknown message type: \(type)")
        }
    }
    
    private func handleASRResult(_ json: [String: Any]) {
        guard let text = json["text"] as? String else { return }
        NotificationCenter.default.post(
            name: .asrResult,
            object: nil,
            userInfo: ["text": text, "isFinal": json["isFinal"] as? Bool ?? true]
        )
    }
    
    private func handleASRInterim(_ json: [String: Any]) {
        guard let text = json["text"] as? String else { return }
        NotificationCenter.default.post(
            name: .asrInterim,
            object: nil,
            userInfo: ["text": text]
        )
    }
    
    private func handleAIReplyChunk(_ json: [String: Any]) {
        guard let text = json["text"] as? String else { return }
        NotificationCenter.default.post(
            name: .aiReplyChunk,
            object: nil,
            userInfo: ["text": text, "isFinal": json["isFinal"] as? Bool ?? false]
        )
    }
    
    private func handleReminder(_ json: [String: Any]) {
        guard let data = json["data"] as? [String: Any] else { return }
        NotificationManager.shared.showLocalNotification(
            title: data["title"] as? String ?? "提醒",
            body: data["body"] as? String ?? ""
        )
    }
    
    private func handleSyncRequired(_ json: [String: Any]) {
        NotificationCenter.default.post(name: .syncRequired, object: nil)
    }
    
    func send(_ message: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: message),
              let text = String(data: data, encoding: .utf8) else {
            return
        }
        webSocket?.send(.string(text)) { error in
            if let error = error {
                print("[WS] Send error: \(error)")
            }
        }
    }
    
    func sendAudioChunk(_ audioData: Data, sequence: Int, isFinal: Bool) {
        let message: [String: Any] = [
            "type": "audio_chunk",
            "data": audioData.base64EncodedString(),
            "sequence": sequence,
            "final": isFinal
        ]
        send(message)
    }
    
    func startASR() {
        send(["type": "asr_start", "format": "pcm", "sampleRate": 16000])
    }
    
    func stopASR() {
        send(["type": "asr_stop"])
    }
    
    private func startPingTimer() {
        pingTimer = Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { [weak self] _ in
            self?.send(["type": "ping"])
        }
    }
    
    private func stopPingTimer() {
        pingTimer?.invalidate()
        pingTimer = nil
    }
}

extension NetworkManager: URLSessionWebSocketDelegate {
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
        DispatchQueue.main.async {
            self.isConnected = true
            self.connectionError = nil
        }
    }
    
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
        DispatchQueue.main.async {
            self.isConnected = false
        }
    }
}

extension NetworkManager {
    func request<T: Decodable>(_ endpoint: String, method: String = "GET", body: [String: Any]? = nil) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(endpoint)") else {
            throw NetworkError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("MASTER", forHTTPHeaderField: "X-User-Role")
        request.setValue(DeviceManager.deviceId, forHTTPHeaderField: "X-Device-Id")
        request.setValue("iOS", forHTTPHeaderField: "X-Platform")
        request.setValue(Bundle.main.appVersion, forHTTPHeaderField: "X-App-Version")
        
        if let body = body {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NetworkError.invalidResponse
        }
        
        guard (200...299).contains(httpResponse.statusCode) else {
            if let error = try? JSONDecoder().decode(APIError.self, from: data) {
                throw NetworkError.apiError(error.error)
            }
            throw NetworkError.httpError(httpResponse.statusCode)
        }
        
        return try JSONDecoder().decode(T.self, from: data)
    }
    
    func chat(message: String, sessionId: String? = nil) async throws -> ChatResponse {
        var body: [String: Any] = ["message": message]
        if let sessionId = sessionId {
            body["sessionId"] = sessionId
        }
        return try await request("/api/conversation/chat", method: "POST", body: body)
    }
}

enum NetworkError: LocalizedError {
    case invalidURL
    case invalidResponse
    case httpError(Int)
    case apiError(String)
    
    var errorDescription: String? {
        switch self {
        case .invalidURL: return "无效的URL"
        case .invalidResponse: return "无效的响应"
        case .httpError(let code): return "HTTP错误: \(code)"
        case .apiError(let message): return message
        }
    }
}

struct APIError: Decodable {
    let error: String
    let code: String?
}

struct ChatResponse: Decodable {
    let reply: String
    let sessionId: String
    let toolCalls: [ToolCall]?
    let intent: String?
    let confidence: Double?
    
    struct ToolCall: Decodable {
        let tool: String
        let args: [String: AnyCodable]?
        let result: [String: AnyCodable]?
    }
}

struct AnyCodable: Codable {
    let value: Any
    
    init(_ value: Any) {
        self.value = value
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let string = try? container.decode(String.self) {
            value = string
        } else if let int = try? container.decode(Int.self) {
            value = int
        } else if let double = try? container.decode(Double.self) {
            value = double
        } else if let bool = try? container.decode(Bool.self) {
            value = bool
        } else if let array = try? container.decode([AnyCodable].self) {
            value = array.map { $0.value }
        } else if let dict = try? container.decode([String: AnyCodable].self) {
            value = dict.mapValues { $0.value }
        } else {
            value = NSNull()
        }
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if let string = value as? String {
            try container.encode(string)
        } else if let int = value as? Int {
            try container.encode(int)
        } else if let double = value as? Double {
            try container.encode(double)
        } else if let bool = value as? Bool {
            try container.encode(bool)
        } else {
            try container.encodeNil()
        }
    }
}

extension Notification.Name {
    static let asrResult = Notification.Name("asrResult")
    static let asrInterim = Notification.Name("asrInterim")
    static let aiReplyChunk = Notification.Name("aiReplyChunk")
    static let syncRequired = Notification.Name("syncRequired")
}

extension Bundle {
    var appVersion: String {
        infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
    }
}
